modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ManagedLocalResult.java

```diff
@@ -43,71 +86,546 @@
      */
     public H2ManagedLocalResult(Session ses, H2MemoryTracker memTracker, Expression[] expressions,
         int visibleColCnt) {
-        super(ses, expressions, visibleColCnt);
+        this.session = ses;
+        rows = Utils.newSmallArrayList();
+        this.visibleColumnCount = visibleColCnt;
+        rowId = -1;
+        this.expressions = expressions;
+        this.memTracker = memTracker;
 
-        this.mem = memTracker;
+        if (ses != null && ses.getQueryContext() != null) {
+            QueryContext qctx = (QueryContext)ses.getQueryContext();
+
+            ctx = qctx.context();
+
+            assert memTracker == null || ctx != null; // Context should be set when we track memory.
+        }
     }
 
-    /** {@inheritDoc} */
-    @Override protected void onUpdate(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
+    /** */
+    private boolean isEnoughMemoryAfterReservation(ValueRow distinctRowKey, Value[] oldRow, Value[] row) {
         assert !isClosed();
         assert row != null;
 
-        long memory;
+        if (memTracker == null)
+            return true; // No memory management set.
+
+        long memory = calculateMemoryDelta(distinctRowKey, oldRow, row);
+
+        memReserved += memory;
+
+        if (memory < 0) {
+            memTracker.released(-memory);
+
+            return true;
+        }
+        else
+            return memTracker.reserved(memory);
+    }
+
+    @Override public boolean isLazy() {
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void setMaxMemoryRows(int maxValue) {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public H2ManagedLocalResult createShallowCopy(SessionInterface targetSession) {
+        if (containsLobs) {
+            return null;
+        }
+
+        ResultExternal e2 = null;
+
+        if (external != null) {
+            e2 = external.createShallowCopy();
+
+            if (e2 == null)
+                return null;
+        }
+
+        H2ManagedLocalResult cp = new H2ManagedLocalResult();
+
+        cp.session = (Session)targetSession;
+        cp.visibleColumnCount = this.visibleColumnCount;
+        cp.expressions = this.expressions;
+        cp.rowId = -1;
+        cp.rowCount = this.rowCount;
+        cp.rows = this.rows;
+        cp.sort = this.sort;
+        cp.distinctRows = this.distinctRows;
+        cp.distinct = distinct;
+        cp.distinctIndexes = distinctIndexes;
+        cp.currentRow = null;
+        cp.offset = 0;
+        cp.limit = -1;
+        cp.containsNull = containsNull;
+        cp.external = e2;
```

Comment:
Do we need to copy memTracker as well?

Reply:
No, because this result is already fully spilled on disk and it doesn't consume memory.